### PR TITLE
clientv3: do not reconnect if request context is canceled

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

**Comment by xiang90** [comment]
LGTM. But probably we should add unit test for this kind of thing?

**Comment by heyitsanthony** [comment]
Sure. I want to avoid testing every retry point though. I think it would suffice to merge the `isRPCError() || ctx.Err()` stuff into a single function and then have one test that invokes a cancelled `Get()`

**Comment by xiang90** [comment]
right. probably having a func called shouldAbort or something like that.

**Comment by heyitsanthony** [comment]
all fixed up /cc @xiang90

File: clientv3/integration/kv_test.go [hunk 1]
```diff
@@ -454,3 +454,27 @@ func TestKVPutFailGetRetry(t *testing.T) {
 	case <-donec:
 	}
 }
+
+// TestKVGetCancel tests that a context cancel on a Get terminates
```
Comment (xiang90) [inline clientv3/integration/kv_test.go]: one line?

File: clientv3/client.go [hunk 2]
```diff
@@ -200,6 +201,9 @@ func dialEndpointList(c *Client) (*grpc.ClientConn, error) {
 	return nil, err
 }
 
-func isRPCError(err error) bool {
-	return strings.HasPrefix(grpc.ErrorDesc(err), "etcdserver: ")
+// isHalted returns true if the given error and context indicate no forward
+// progress can be made, even after reconnecting.
+func isHalted(ctx context.Context, err error) bool {
```
Comment (xiang90) [inline clientv3/client.go]: add a unit test for isHalted?

**Comment by xiang90** [comment]
lgtm

**Pull request merged**